### ISSUE TITLE
Change root path in chaos-mesh dashboard

### ISSUE THREAD

**Issue by nml11c** [chaos-mesh/chaos-mesh#2697]
## Question

I want to host dashboard under `<common-fqdn>/chaos-mesh`, instead of `<unique-fqdn-for-chaos-mesh>/`. Is it possible to achieve?

**Comment by YangKeao** [comment]
As I know, the dashboard frontend cannot work on a subpath. @g1eny0ung Am I correct?

**Comment by g1eny0ung** [comment]
@YangKeao Can this be solved by configuring ingress?

https://github.com/chaos-mesh/chaos-mesh/blob/master/helm/chaos-mesh/values.yaml#L298

But I haven't tried this configuration.

**Comment by YangKeao** [comment]
@g1eny0ung Will the frontend work? (considering the frontend route or something).

I think the backend is never a problem (there are tons of ways to rewrite/proxy the http request.)

**Comment by g1eny0ung** [comment]
> @g1eny0ung Will the frontend work? (considering the frontend route or something). I will have a try later.
> 
> I think the backend is never a problem (there are tons of ways to rewrite/proxy the http request.)

Yep. The backend request is easy to rewrite. The key to the problem is front-end routing.

The frontend will redirect the root path to `/dashboard`, so I think if you access `/dashboard` directly there will be no problem.

**Comment by YangKeao** [comment]
@g1eny0ung @nml11c 

I have tested. It doesn't work! The first problem is the resource link (`/static/js/main.cce2d46a.chunk.js`, `/static/js/2.396a6d11.chunk.js`...), which is coded with absolute path.

**Comment by g1eny0ung** [comment]
> @g1eny0ung @nml11c
> 
> I have tested. It doesn't work! The first problem is the resource link (`/static/js/main.cce2d46a.chunk.js`, `/static/js/2.396a6d11.chunk.js`...), which is coded with absolute path.

🤔 Really need to modify some paths and front-end redirects. I think I can do this and test.

**Comment by nmlc** [comment]
Am I correct to assume, that you guys recognize this as an usability issue and going to fix it?

**Comment by g1eny0ung** [comment]
It will be fixed. Thank you for your continued tracking, and we realize that this is a usability issue. ❤️

**Comment by kisahm** [comment]
is there a fix already available?

**Comment by YangKeao** [comment]
@g1eny0ung Do you have any plan/timeline to fix this issue?

**Comment by andyzimmerman1** [comment]
@g1eny0ung I am running into this issue as well. It looks like it has been several months since the last update - do you know when a fix will hit the baseline?

**Comment by f4tal-err0r** [comment]
Also running into this issue. As a workaround im temporarily rewriting this but it is very hacky and suboptimal.

**Comment by wujunfu** [comment]
Hi @YangKeao , is there any updates on this issue?

**Comment by g1eny0ung** [comment]
Sorry guys, I've ignored this usage issue for a long time and I will try to fix it within this week.

**Comment by g1eny0ung** [comment]
Hi guys, after testing all afternoon, I finally managed to deploy the chaos dashboard on a subpath. 😵 And here are some details to clarify:

To minimize changes, you don't have to change the existing configuration of the dashboard. But it doesn't come without a price. The price is **now the frontend router will be a hash router rather than a browser router**. Here is the detailed documentation on both of them:

- https://reactrouter.com/en/main/routers/create-browser-router
- https://reactrouter.com/en/main/routers/create-hash-router

As an example, what used to be `/workflows` will become `#/workflows`.

I have tried many ways to keep the old browser router approach, but it is impossible to do so. Because chaos dashboard is a single-page application, the backend can only redirect all requests to `index.html` and then let the frontend manage how to do the jumping itself.

Here's the problem, when I click a link on the front end to make a jump, I can't dynamically add the specified `subpath` which I defined in the ingress, it is [a configurable item](https://reactrouter.com/en/main/router-components/browser-router#basename) that has to be packaged into the frontend static files by way of compilation.

I figured out some other dashboards like the Kubernetes dashboard, It does also use a hash router to make it easy for users to access it as an arbitrary proxy. So I have also convinced myself that it is feasible to adopt this approach. 😉

In summary, If #4093 is finally merged, **the dashboard will have a breaking change, i.e. any path will have to have a `#` prefix before it can be accessed.**

I have also added an example of nginx ingress as a reference: https://github.com/chaos-mesh/chaos-mesh/pull/4093/files#diff-23b5425b9740919d17940e94316e228bee63c26b1dd3683b17a3fd93c771153d.

If you're interested, you can switch to the related PR to help test. Thank you all for your continued interest in this long existed issue. ❤️

**Comment by kranthikirang** [comment]
@g1eny0ung I am not an expert in react but I have achieved subdirectory hosting in react by following https://stackoverflow.com/questions/37396427/how-to-bundle-a-react-app-to-a-subdirectory-on-a-server 

also nginx rewrite with $2; This is not document root change this is hosting under subpath(sub directory). So that react will send /subdirectory appended to the URI. 

this is my approuter example

`import React from "react";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import Menu from "../Components/Common/Menu";
import Home from "../Components/Home";
import Health from "../Components/Health";
import About from "../Components/Page/About";
import Admin from "../Components/Admin";
import Contact from "../Components/Page/Contact";
import NoPage from "../Components/Page/NoPage";

function AppRouter() {
  return (
    <Router basename={process.env.PUBLIC_URL}>
        <Menu />
        
        <div>
            <Routes>
                <Route path={'/'} element={<Home />} />
                <Route path={'/health'} element={<Health />} />
                <Route path={'/about'} element={<About />} />
                <Route path={'/contact'} element={<Contact />} />
                <Route path="*" element={<NoPage />} />
            </Routes>
        </div>
    </Router>
  );
}

export default AppRouter;`

So, while building the image for behind nginx proxy all I do is supply myreayapp subpath

PUBLIC_URL=/myreactapp npm run build

for normal hosting

PUBLIC_URL=/ npm run build

So, that we can avoid hash router which may breakthings

**Comment by g1eny0ung** [comment]
@kranthikirang Yep. Using `PUBLIC_URL` can specify the subpath. But it can only happen in build time. We should build an out-of-box image rather than let users build it every time for themselves.

**Comment by kranthikirang** [comment]
probably hashrouter is the only way to go then
